src/form_beam.cpp

```diff
@@ -248,128 +256,101 @@ __global__ void vmBeamform_kernel( gpuDoubleComplex *Jv_Q,
+     * and this is the number of tiles necessary to cover the entire input.
      */
-    gpuDoubleComplex *ex  = (gpuDoubleComplex *)(&arrays[0*nant]);
-    gpuDoubleComplex *ey  = (gpuDoubleComplex *)(&arrays[2*nant]);
-    gpuDoubleComplex *Nxx = (gpuDoubleComplex *)(&arrays[4*nant]);
-    gpuDoubleComplex *Nxy = (gpuDoubleComplex *)(&arrays[6*nant]);
-    gpuDoubleComplex *Nyy = (gpuDoubleComplex *)(&arrays[8*nant]);
-    // (Nyx is not needed as it's degenerate with Nxy)
-
-    // Calculate the beam and the noise floor
-    /* Fix from Maceij regarding NaNs in output when running on Athena, 13 April 2018.
-    Apparently the different compilers and architectures are treating what were
-    unintialised variables very differently */
-    ex[ant]  = make_gpuDoubleComplex( 0.0, 0.0 );
-    ey[ant]  = make_gpuDoubleComplex( 0.0, 0.0 );
-
-    Nxx[ant] = make_gpuDoubleComplex( 0.0, 0.0 );
-    Nxy[ant] = make_gpuDoubleComplex( 0.0, 0.0 );
-    Nyy[ant] = make_gpuDoubleComplex( 0.0, 0.0 );
-    __syncthreads();
-
-    // Calculate beamform products for each antenna, and then add them together
-    // Calculate the coherent beam (B = J*phi*D)
-    ex[ant] = gpuCmul( phi[PHI_IDX(p,ant,c,nant,nc)], Jv_Q[Jv_IDX(p,s,c,ant,ns,nc,nant)] );
-    ey[ant] = gpuCmul( phi[PHI_IDX(p,ant,c,nant,nc)], Jv_P[Jv_IDX(p,s,c,ant,ns,nc,nant)] );
-
-    Nxx[ant] = gpuCmul( ex[ant], gpuConj(ex[ant]) );
-    Nxy[ant] = gpuCmul( ex[ant], gpuConj(ey[ant]) );
-    Nyy[ant] = gpuCmul( ey[ant], gpuConj(ey[ant]) );
-    __syncthreads();
-
-    // Detect the coherent beam
-    // The safest, slowest option: Just get one thread to do it
-    if ( ant == 0 )
-    {
-        for (int i = 1; i < nant; i++)
-        {
-            ex[0]  = gpuCadd( ex[0],  ex[i] );
-            ey[0]  = gpuCadd( ey[0],  ey[i] );
-            Nxx[0] = gpuCadd( Nxx[0], Nxx[i] );
-            Nxy[0] = gpuCadd( Nxy[0], Nxy[i] );
-            Nyy[0] = gpuCadd( Nyy[0], Nyy[i] );
-       }
-    }
-    __syncthreads();
-
-    // Form the stokes parameters for the coherent beam
-    // Only doing it for ant 0 so that it only prints once
-    if ( ant == 0 )
-    {
-        float bnXX = DETECT(ex[0]) - gpuCreal(Nxx[0]);
-        float bnYY = DETECT(ey[0]) - gpuCreal(Nyy[0]);
-        gpuDoubleComplex bnXY = gpuCsub( gpuCmul( ex[0], gpuConj( ey[0] ) ),
-                                    Nxy[0] );
-
-        // Stokes I, Q, U, V:
-        S[C_IDX(p,s+soffset,0,c,ns*nchunk,nstokes,nc)] = invw*(bnXX + bnYY);
-        if ( nstokes == 4 )
-        {
-            S[C_IDX(p,s+soffset,1,c,ns*nchunk,nstokes,nc)] = invw*(bnXX - bnYY);
-            S[C_IDX(p,s+soffset,2,c,ns*nchunk,nstokes,nc)] =  2.0*invw*gpuCreal( bnXY );
-            S[C_IDX(p,s+soffset,3,c,ns*nchunk,nstokes,nc)] = -2.0*invw*gpuCimag( bnXY );
+    const unsigned int n_loops {(n_samples * nfine_chan + total_warps - 1) / total_warps};
+
+    __shared__ gpuDoubleComplex workspace[NTHREADS_BEAMFORMING_KERNEL * 5];
```

Comment:
I notice here and through that there is a multiplication by `5` - why? What is the significance of `5` rather than `7`, for example? Or is it because there are 5 "values" we keep in shared memory (ex, ey, Nxx, Nyy, Nxy)? 

Is it something that should be codified as a variable somewhere to use, so that if it ever needs to be changed it's a simple one-line change?

Reply:
> Or is it because there are 5 "values" we keep in shared memory (ex, ey, Nxx, Nyy, Nxy)?

This is it, 5 different variables for each thread.

Reply:
Right, okay. Perhaps it is a good idea to explicitly note that when the number first arises?

Reply:
Done.